amazon/ion/ioncmodule.c

```diff
@@ -1275,19 +1275,27 @@ iERR ionc_read_value(hREADER hreader, ION_TYPE t, PyObject* container, BOOL in_s
         }
         case tid_STRUCT_INT:
         {
-            ion_nature_constructor = _ionpydict_fromvalue;
-            //Init a IonPyDict
-            PyObject* new_dict = PyDict_New();
+            PyObject* data = PyDict_New();
+            IONCHECK(ionc_read_into_container(hreader, data, /*is_struct=*/TRUE, emit_bare_values));
+
             py_value = PyObject_CallFunctionObjArgs(
-                ion_nature_constructor,
-                py_ion_type_table[ion_type >> 8],
-                new_dict,
-                py_annotations,
-                NULL
+
+                    _ionpydict_factory,
+                    data,
+                    py_annotations,
+                    NULL
             );
-            Py_XDECREF(new_dict);
+            if (py_value == NULL) {
+                printf("struct py value is null");
+                PyErr_Print();
```

Comment:
Looks like this needs to be cleaned up?

Reply:
The `printf` and `PyErr_Print`: definitely. But in theory we should _always_ be checking the return values from these calls to the Python runtime and setting/exposing helpful error codes, and we're not anywhere else.

Concretely, if as a developer/tinkerer, I change the name or signature of the python function then this will return NULL and set a PyErr code. Without error checking it's pretty hard to track down what the problem is. I don't necessarily want to add error checking everywhere else, but it would feel odd to have this be the one place we check for NULL and then set an appropriate PyErr to have an exception propagate.

Reply:
I removed the prints in favor of the existing IERR pattern.

I also created https://github.com/amazon-ion/ion-python/issues/310 to track propagating Python Error information out with the exception that is thrown.